testing/trino-tests/src/test/java/io/trino/security/TestImpersonation.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.security;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Key;
+import io.airlift.log.Logging;
+import io.trino.jdbc.TrinoConnection;
+import io.trino.metadata.SystemSecurityMetadata;
+import io.trino.plugin.memory.MemoryPlugin;
+import io.trino.server.testing.TestingTrinoServer;
+import io.trino.spi.security.TrinoPrincipal;
+import io.trino.testing.TestingAccessControlManager;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.parallel.Execution;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Optional;
+
+import static com.google.inject.multibindings.OptionalBinder.newOptionalBinder;
+import static io.trino.spi.security.PrincipalType.USER;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+public class TestImpersonation
```

Comment:
Should we move this tests to `trino-client` or `trino-jdbc`?

Reply:
That would be great but we can't due to the necessary dependencies. 
You don't want trino-jdbc to depend on trino-main.

Reply:
I still believe it would be weird for `trino-main` to depend on `trino-client` or `trino-jdbc`

Reply:
> You don't want trino-jdbc to depend on trino-main.

And 

>  it would be weird for trino-main to depend on trino-client or trino-jdbc

are cyclic. I am getting lost.

I think it is ok, for trino-client or trino-jdbc to depend on trino-main in test scope because you need to have a server to test the client.

Can you please explain to me why it is weird? I would be afraid of cyclic dependencies which may happen.

Reply:
Ok but then `trino-client` or `trino-jdbc` would still need a dependency on `trino-tests` because I rely on `TestingSystemSecurityMetadata`. And I don't think we want to rewrite/copy this class